File: 2023/day1/part1.py
def get_calibration_value(line):
  text_digits = ['0', '1', '2', '3', '4', '5', '6', '7', '8', '9']
  text_value = ''
  for c in line:
    if c in text_digits:
      text_value += c
      break
  for c in line[::-1]:
    if c in text_digits:
      text_value += c
      break
  value = int(text_value)
  return value


def main():
  tot = 0
  with open('input.txt') as f:
    for line in f:
      val = get_calibration_value(line)
      tot += val
  print(tot)

if __name__ == '__main__':
  main()

File: 2023/day12/part1.nim
import std/strutils
import std/sequtils

type
  Spring* = enum
    Unknown, Damaged, Operational

proc parseSpring*(c: char): Spring =
  case c
  of '?': result = Unknown
  of '#': result = Damaged
  of '.': result = Operational
  else: echo "Should not happen"

proc count_arrangements(springs: var seq[Spring], damaged: sink seq[int], idx: sink int = 0, damagedCount: sink int = 0): int = 
  while idx < springs.len:
    case springs[idx]
    of Unknown:
      springs[idx] = Damaged
      let cntDmg = count_arrangements(springs, damaged, idx, damagedCount)
      springs[idx] = Operational
      let cntoP = count_arrangements(springs, damaged, idx, damagedCount)
      springs[idx] = Unknown
      return cntDmg + cntOp
    of Damaged:
      inc damagedCount
      if damaged.len == 0 or damagedCount > damaged[0]:
        return 0
    of Operational:
      if damagedCount == 0:
        discard
      elif damagedCount == damaged[0]:
        damaged.delete(0)
        damagedCount = 0
      else:
        return 0
    inc idx
  if damaged.len == 0:
    return 1
  elif damaged.len == 1 and damagedCount == damaged[0]:
    return 1
  else:
    return 0


proc main() =
  var tot = 0
  for line in lines("input.txt"):
    let s = line.split(' ')
    var springs = s[0].map(parseSpring)
    let damaged = s[1].split(',').map(parseInt)
    tot += count_arrangements(springs, damaged)
  echo tot

when isMainModule:
  main()